### PR TITLE
repos: Remove use of advisory lock

### PR DESCRIPTION
**Pull request by ryanslade**
Repo syncing can only happen in repo-updater, of which we only ever run
one instance. We can avoid using advisory locks by instead using the
singleflight package to ensure that only one sync per repo can happen
concurrently.

As part of the change `syncRepo` now always blocks and it is up to `SyncRepo`
to decide whether to spawn a goroutine or not.

This change was motivated by performance issues on sourcegraph.com where
we observed repo-updater blocking for up to three minutes waiting on
advisory locks which would in turn starve it of database connections.

Part of https://github.com/sourcegraph/sourcegraph/issues/25153

### CODE REVIEW

**Review by tsenart — APPROVED** [review]
LGTM, hope this fixes it!

File: internal/repos/syncer.go [hunk 1]
```diff
@@ -262,20 +265,37 @@ func (s *Syncer) SyncRepo(ctx context.Context, name api.RepoName) (repo *types.R
 	}
 
 	if repo == nil {
-		// We don't have this repo yet, so block before returning.
-		return s.syncRepo(ctx, codehost, name, nil)
+		// We don't have this repo yet so we need to try and sync it now and we will block
+		// until syncing is complete
+		repo, err, _ := s.syncGroup.Do(string(name), func() (interface{}, error) {
```
Comment (indradhanush) [inline internal/repos/syncer.go:270]: Do we want to ignore the `shared` bool returned from `s.syncGroup.Do`? :thinking: 
I'm wondering if we should be logging it as an error if `shared` is `true`? :thinking:

File: internal/repos/syncer.go [hunk 2]
```diff
@@ -262,20 +265,37 @@ func (s *Syncer) SyncRepo(ctx context.Context, name api.RepoName) (repo *types.R
...
 	// Sync the repo in the background if it wasn't updated in 1 minute.
 	if s.Now().Sub(repo.UpdatedAt) >= time.Minute {
 		go func() {
 			ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
 			defer cancel()
-			s.syncRepo(ctx, codehost, name, repo)
+
+			// We don't care about the return value here, but we still want to ensure that
+			// only one is in flight at a time.
+			_, _, _ = s.syncGroup.Do(string(name), func() (interface{}, error) {
```
Comment (indradhanush) [inline internal/repos/syncer.go:292]: Maybe we also want to check the value of `shared` here based on the comment?

File: internal/repos/syncer.go [hunk 3]
```diff
@@ -262,20 +265,37 @@ func (s *Syncer) SyncRepo(ctx context.Context, name api.RepoName) (repo *types.R
 	}
 
 	if repo == nil {
-		// We don't have this repo yet, so block before returning.
-		return s.syncRepo(ctx, codehost, name, nil)
+		// We don't have this repo yet so we need to try and sync it now and we will block
+		// until syncing is complete
```
Comment (indradhanush) [inline internal/repos/syncer.go:269]:
```suggestion
		// until syncing is complete.
```

:stuck_out_tongue_closed_eyes:

**Review by indradhanush — APPROVED** [review]
LGTM. Approve to unblock.

I like using this `singleflight` package - even though I've not used it previously I think this is a very elegant solution. :ok_hand: 

My only thought around not ignoring the `shared` variable returned from the `Do` method is that if in case we have underlying issues with repo syncing sometime in the future, the errors that we log if `shared == true` could help us spot the issues sooner rather than waiting on something major to happen. 

:crossed_fingers: this fixes the issue.

File: internal/repos/syncer.go (hunk 2 above)
Comment (ryanslade) [reply internal/repos/syncer.go:292]: `true` here just means that more than one routine actually got the result. I don't think that's an issue for us. 

Also, we don't have access to shared until after the outer `Do` completes. If we log there then we'll get a log line for each routine that used the value which is probably not what we want.

**Pull request merged**